src/smtlib_reader.cpp

```diff
@@ -0,0 +1,278 @@
+}
+
+Result SmtLibReader::check_sat()
+{
+  Result r = solver_->check_sat();
+  cout << r << endl;
+  return r;
+}
+
+Result SmtLibReader::check_sat_assuming(const TermVec & assumptions)
+{
+  Result r = solver_->check_sat_assuming(assumptions);
+  cout << r << endl;
+  return r;
+}
+
+void SmtLibReader::push(uint64_t num)
+{
+  for (size_t i = 0; i < num; ++i)
+  {
+    global_symbols_.push_scope();
+  }
+  solver_->push(num);
+}
+
+void SmtLibReader::pop(uint64_t num)
+{
+  for (size_t i = 0; i < num; ++i)
+  {
+    global_symbols_.pop_scope();
+  }
+  solver_->pop(num);
+}
+
+void SmtLibReader::push_scope()
+{
+  arg_param_map_.push_scope();
+  sort_arg_ids_.push_back(std::unordered_map<smt::Sort, size_t>());
+}
+
+void SmtLibReader::pop_scope()
+{
+  arg_param_map_.pop_scope();
+  sort_arg_ids_.pop_back();
+}
+
+Term SmtLibReader::lookup_symbol(const string & sym)
+{
+  Term symbol_term;
+  assert(!symbol_term);
+
+  if (current_scope())
+  {
+    // check scoped variables before global symbols
+    // shadowing semantics
+    try
+    {
+      return arg_param_map_.get_symbol(sym);
+    }
+    catch (std::out_of_range & e)
+    {
+      ;
+    }
+  }
+
+  assert(!symbol_term);
+  try
+  {
+    return global_symbols_.get_symbol(sym);
+  }
+  catch (std::out_of_range & e)
+  {
+    ;
+  }
+  return symbol_term;
+}
+
+void SmtLibReader::new_symbol(const std::string & name, const smt::Sort & sort)
+{
+  try
+  {
+    Term t = global_symbols_.get_symbol(name);
+    throw SmtException("Re-declaring symbol: " + name);
+  }
+  catch (std::out_of_range & e)
+  {
+    auto it = all_symbols_.find(name);
+    if (it != all_symbols_.end())
+    {
+      if (it->second->get_sort() != sort)
+      {
+        throw SmtException("Current Limitation: cannot re-declare symbol "
+                           + name + " with a different sort");
+      }
+      global_symbols_.add_mapping(name, it->second);
+      return;
+    }
+  }
+
+  Term fresh_symbol = solver_->make_symbol(name, sort);
+  global_symbols_.add_mapping(name, fresh_symbol);
+  all_symbols_[name] = fresh_symbol;
+}
+
+PrimOp SmtLibReader::lookup_primop(const std::string & str)
+{
+  return str2primop.at(str);
+}
+
+void SmtLibReader::define_fun(const string & name,
+                              const Term & def,
+                              const TermVec & args)
+{
+  if (args.size())
+  {
+    // this is a function
+    defs_[name] = def;
+    def_args_[name] = args;
+  }
+  else
+  {
+    // this is just an alias for another term
+    global_symbols_.add_mapping(name, def);
+  }
+}
+
+Term SmtLibReader::apply_define_fun(const string & defname,
+                                    const TermVec & args)
+{
+  UnorderedTermMap subs_map;
+  size_t num_args = args.size();
+
+  auto it = defs_.find(defname);
+
+  if (it == defs_.end())
+  {
+    throw SmtException("Unknown function: " + defname);
+  }
+
+  Term def = it->second;
+
+  if (!num_args)
+  {
+    def;
```

Comment:
Lol what is happening there? Thanks for catching that! Meant to be a return, fixed with e7e7531eb20ecb3d74762b964f7b34192a188aff. Not an actual bug because all the code afterwards still works, but if there are no arguments then we don't need to substitute at all.

Reply:
And actually, on second thought that code should never run at all. Because `apply_define_fun` should only get run with arguments. I'll change that into an assertion.